### ISSUE TITLE
crypto_kx_client/server_session_keys

### ISSUE THREAD

**Issue by fredrikzetterman** [jedisct1/libsodium#536]
Hi,

I'm a crypto newbie and I've just started using libsodium. I have a question regarding crypto_kx_client_session_keys/crypto_kx_server_session_keys, reading about them here: https://download.libsodium.org/doc/key_exchange/

Everything on that page is crystal clear, _except_ that I don't see how I should use the generated session keys. If I 'grep' after crypto_kx_SESSIONKEYBYTES I only get a hit in crypto_kx.h. Are the keys safe to use with crypto_secretbox_easy() for example, since crypto_kx_SESSIONKEYBYTES == crypto_secretbox_KEYBYTES?

**Comment by BurningEnlightenment** [comment]
_Generally_ it isn't safe to use the result of a DH/ECDH key exchange as key material for symmetric encryption functions, because the values are not uniformly distributed (pseudo-)random strings. I.e. an additional key derivation step would be necessary.
However, the algorithm details section for the `crypto_kx_*` says that the output of the ECDH key exchange is hashed using blake2b-512 which is a PRF and generally an acceptable KDF. So you can safely use the output of the `crypto_kx_*` functions as keys for the `crypto_secretbox_easy` API. If you need more key material and/or bind the keys to specific contexts, you can use [`crypto_kdf`](https://download.libsodium.org/doc/key_derivation/) to stretch the high entropy key material output by `crypto_kx_*`

**TLDR**; The algorithm details section says that the output of the ECDH key exchange is hashed using blake2b-512, so it is safe to use them as input for crypto_secretbox_easy.

**Comment by jedisct1** [comment]
Excellent question, and excellent response from @BurningEnlightenment , thanks!

And a good heads up for me that I should update the documentation to clarify this.

The output can indeed be used with anything requiring a key which is 256-bit or less (truncating these keys is fine, but with the exception of SipHash, all the keys in libsodium are 256-bit).

**Comment by fredrikzetterman** [comment]
Thank you for the prompt replies, good information and a superb API to work with!
